sale_delivery_carrier_preference/models/sale_order.py

```diff
@@ -1,26 +1,43 @@
 # Copyright 2020 Camptocamp SA
 # License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl)
 
-from odoo import api, fields, models
+from odoo import _, api, exceptions, fields, models
 
 
 class SaleOrder(models.Model):
 
     _inherit = "sale.order"
 
     shipping_weight = fields.Float(
-        string="Shipping weight (kg)", compute="_compute_shipping_weight",
+        string="Shipping weight (kg)", compute="_compute_shipping_weight"
     )
 
     def action_confirm(self):
+        self._add_delivery_carrier_on_confirmation()
+        return super().action_confirm()
+
+    def _add_delivery_carrier_on_confirmation(self):
         """Automatically add delivery.carrier on sale order confirmation"""
         for order in self:
-            if (
-                any([line.product_id.type != "service" for line in order.order_line])
-                and not order.carrier_id
-            ):
-                order.carrier_id = order.get_preferred_carrier()
-        return super().action_confirm()
+            if order.carrier_id or any(line.is_delivery for line in order.order_line):
+                continue
+            carrier = order.get_preferred_carrier()
+            if not carrier:
+                continue
+            vals = carrier.rate_shipment(order)
+            if vals.get("success"):
+                delivery_price = vals["price"]
+                order.set_delivery_line(carrier, delivery_price)
+                order.recompute_delivery_price = False
+                if vals.get("warning_message"):
+                    order.delivery_message = vals["warning_message"]
+            else:
+                raise exceptions.UserError(
```

Comment:
This will restrict usage of any `delivery.carrier` that doesn't have a rate integration. I'm not sure what must be done in such case but I'm not sure preventing the confirmation is right.

Reply:
Now it will raise an error only when success is explicitly set to False in a dict returned by `rate_shipment`.